compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -459,13 +457,184 @@ struct DistributeBroadcast final : OpDistributionPattern<vector::BroadcastOp> {
+    if (layout.getThreadActiveIds()[i]) {
+      activeThreadStrides.push_back(stride);
+    }
+  }
+  // Apply permutation on the active thread strides.
+  // TODO: Do we need to do inversion or not?
+  applyPermutationToVector(activeThreadStrides, layout.getThreadOrder());
+  return activeThreadStrides[dim];
+}
+
+static int64_t getShuffleWidth(NestedLayoutAttr layout, int64_t dim) {
+  return layout.getThreadsPerOuter()[dim];
+}
+
+struct DistributeMultiReduction final
+    : OpDistributionPattern<vector::MultiDimReductionOp> {
+  using OpDistributionPattern::OpDistributionPattern;
+
+  DistributeMultiReduction(MLIRContext *context, int64_t maxBitsPerShuffle,
+                           int64_t benefit = 1)
+      : OpDistributionPattern(context, benefit),
+        maxBitsPerShuffle(maxBitsPerShuffle) {}
+
+  LogicalResult matchAndRewrite(vector::MultiDimReductionOp multiReduceOp,
+                                DistributionSignature &signature,
+                                PatternRewriter &rewriter) const override {
+    VectorValue srcVector = multiReduceOp.getSource();
+    VectorValue accVector = dyn_cast<VectorValue>(multiReduceOp.getAcc());
+    if (!accVector) {
+      return rewriter.notifyMatchFailure(
+          multiReduceOp, "unimplemented: scalar accumulator distribution");
+    }
+    VectorValue resVector = dyn_cast<VectorValue>(multiReduceOp.getResult());
+    if (!resVector) {
+      return rewriter.notifyMatchFailure(
+          multiReduceOp, "unimplemented: scalar result distribution");
+    }
+
+    auto srcLayout = dyn_cast_or_null<NestedLayoutAttr>(signature[srcVector]);
+    if (!srcLayout) {
+      return rewriter.notifyMatchFailure(multiReduceOp,
+                                         "expected nested layout attr");
+    }
+
+    VectorValue disSrc =
+        getDistributed(rewriter, srcVector, signature[srcVector]);
+    VectorValue disAcc =
+        getDistributed(rewriter, accVector, signature[accVector]);
+
+    Location loc = multiReduceOp.getLoc();
+
+    SmallVector<bool> reducedDims = multiReduceOp.getReductionMask();
+    int64_t rank = srcVector.getType().getRank();
+
+    // Do thread local reduce.
+
+    SmallVector<bool> distributedReductionMask;
+    distributedReductionMask.reserve(3 * rank);
+    distributedReductionMask.append(
+        applyPermutation(reducedDims, srcLayout.getBatchOrder()));
+    distributedReductionMask.append(
+        applyPermutation(reducedDims, srcLayout.getOuterOrder()));
+    distributedReductionMask.append(
+        applyPermutation(reducedDims, srcLayout.getElementOrder()));
+
+    auto localReduction = rewriter.create<vector::MultiDimReductionOp>(
+        loc, disSrc, disAcc, distributedReductionMask, multiReduceOp.getKind());
+    auto locallyReduced = dyn_cast<VectorValue>(localReduction.getResult());
+
+    if (!locallyReduced) {
+      return rewriter.notifyMatchFailure(
+          localReduction, "unimplemented: scalar result distribution");
```

Comment:
`MultiDimReduction` returns a scalar if and only if not all of the dimensions are reduced, and that is already guaranteed with the earlier check so this can just be an assert. Best practice is to avoid pattern failures after generating IR also.

https://mlir.llvm.org/docs/Dialects/Vector/#vectormulti_reduction-vectormultidimreductionop